### PR TITLE
Allow filesystem commands to access files with glob metachars in name

### LINKED ISSUE #10571: `cp`, `ls` and others cannot access file that is actually present (and accessible).
### Describe the bug

 Apropos the discussion at the bottom of #10557 

If a file path contains glob metachars, you cannot access it from any command that takes arguemnts that are globs (unless you do extra glob quoting).

### How to reproduce

Example:
```
> ls **/*
╭───┬───────┬──────┬──────┬────────────────╮
│ # │ name  │ type │ size │    modified    │
├───┼───────┼──────┼──────┼────────────────┤
│ 0 │ a[b]d │ file │  0 B │ 35 seconds ago │
│ 1 │ foo   │ dir  │  2 B │ a minute ago   │
╰───┴───────┴──────┴──────┴────────────────╯
> glob a[b]d
╭────────────╮
│ empty list │
╰────────────╯
> ucp a[b]d foo
Error: nu::shell::file_not_found

  × File not found
   ╭─[entry #41:1:1]
 1 │ ucp a[b]d foo
   ·     ──┬──
   ·       ╰── file not found
   ╰────

> cp a[b]d foo
Error:   × No matches found
   ╭─[entry #42:1:1]
 1 │ cp a[b]d foo
   ·    ──┬──
   ·      ╰── no matches found
   ╰────

> ls a[b]d
Error:   × No matches found for a[b]d
   ╭─[entry #44:1:1]
 1 │ ls a[b]d
   ·    ──┬──
   ·      ╰── Pattern, file or folder not found
   ╰────
  help: no matches found

```
The file exists, but cannot be accessed by any of the above command.

You can access it, but need to resort to quoting in the command argument:

```
> ls a[b]d
Error:   × No matches found for a[b]d
   ╭─[entry #44:1:1]
 1 │ ls a[b]d
   ·    ──┬──
   ·      ╰── Pattern, file or folder not found
   ╰────
  help: no matches found

> ls a[[]b[]]d
╭───┬───────┬──────┬──────┬────────────────╮
│ # │ name  │ type │ size │    modified    │
├───┼───────┼──────┼──────┼────────────────┤
│ 0 │ a[b]d │ file │  0 B │ 14 minutes ago │
╰───┴───────┴──────┴──────┴────────────────╯

```

### Expected behavior

I think user is right to expect `ls`, `cp` to be able to access the file whose name is explicitly provided.  The path contains unusual, but perfectly legal chars.

Further, the user is not likely to try glob quoting with `[]` to workaround this issue.  S/he is more likely to try string quoting with `\` and get frustrated that does not work.

### Screenshots

_No response_

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.85.1                                   |
| branch             | wax_glob2                                |
| commit_hash        | 9ced0359c89029bcbc2a4c251d170c5be797200f |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.70.0 (90c541806 2023-05-31)      |
| rust_channel       | 1.70.0-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.70.0 (ec8a8a0ca 2023-04-25)      |
| build_time         | 2023-10-01 10:13:49 -07:00               |
| build_rust_channel | debug                                    |
| allocator          | mimalloc                                 |
| features           | default, sqlite, trash, which, zip       |
| installed_plugins  |                                          |

### Additional context

Note that we have this problem with glob characters that were supported before PR #10557, and with commands that use both the new and the old glob algorithm.

What to do?  For all of the commands that take a glob as argument, with the possible exception of `glob` command, I think the right thing to do is first try the arg as a literal path (e.g `stat` the arg), only proceed to try it as a glob if the `stat` fails.

### PR DESCRIPTION
**Pull request by bobhy**
(squashed version of #10557, clean commit history and review thread)

Fixes #10571, also potentially: #10364, #10211, #9558, #9310,

# Description
Changes processing of arguments to filesystem commands that are source paths or globs.
Applies to `cp, cp-old, mv, rm, du` but not `ls` (because it uses a different globbing interface) or `glob` (because it uses a different globbing library).

The core of the change is to lookup the argument first as a file and only glob if it is not.  That way,
a path containing glob metacharacters can be referenced without glob quoting, though it will have to be single quoted to avoid nushell parsing.

Before: A file path that looks like a glob is not matched by the glob specified as a (source) argument and takes some thinking about to access. You might say the glob pattern shadows a file with the same spelling.
```
> ls a*
╭───┬────────┬──────┬──────┬────────────────╮
│ # │  name  │ type │ size │    modified    │
├───┼────────┼──────┼──────┼────────────────┤
│ 0 │ a[bc]d │ file │  0 B │ 34 seconds ago │
│ 1 │ abd    │ file │  0 B │ now            │
│ 2 │ acd    │ file │  0 B │ now            │
╰───┴────────┴──────┴──────┴────────────────╯

> cp --verbose 'a[bc]d' dest
copied /home/bobhy/src/rust/work/r4/abd to /home/bobhy/src/rust/work/r4/dest/abd
copied /home/bobhy/src/rust/work/r4/acd to /home/bobhy/src/rust/work/r4/dest/acd

> ## Note -- a[bc]d *not* copied, and seemingly hard to access.
> cp --verbose 'a\[bc\]d' dest
Error:   × No matches found
   ╭─[entry #33:1:1]
 1 │ cp --verbose 'a\[bc\]d' dest
   ·              ─────┬────
   ·                   ╰── no matches found
   ╰────

> #.. but is accessible with enough glob quoting.
> cp --verbose 'a[[]bc[]]d' dest
copied /home/bobhy/src/rust/work/r4/a[bc]d to /home/bobhy/src/rust/work/r4/dest/a[bc]d
```
Before_2: if file has glob metachars but isn't a valid pattern, user gets a confusing error:

```
> touch 'a[b'
> cp 'a[b' dest
Error:   × Pattern syntax error near position 30: invalid range pattern
   ╭─[entry #13:1:1]
 1 │ cp 'a[b' dest
   ·    ──┬──
   ·      ╰── invalid pattern
   ╰────
```

After: Args to cp, mv, etc. are tried first as literal files, and only as globs if not found to be files.

```
> cp --verbose 'a[bc]d' dest
copied /home/bobhy/src/rust/work/r4/a[bc]d to /home/bobhy/src/rust/work/r4/dest/a[bc]d
> cp --verbose '[a][bc]d' dest
copied /home/bobhy/src/rust/work/r4/abd to /home/bobhy/src/rust/work/r4/dest/abd
copied /home/bobhy/src/rust/work/r4/acd to /home/bobhy/src/rust/work/r4/dest/acd
```
After_2: file with glob metachars but invalid pattern just works. (though Windows does not allow file name to contain `*`.).

```
> cp --verbose 'a[b' dest
copied /home/bobhy/src/rust/work/r4/a[b to /home/bobhy/src/rust/work/r4/dest/a[b
```

So, with this fix, a file shadows a glob pattern with the same spelling. If you have such a file and really want to use the glob pattern, you will have to glob quote some of the characters in the pattern.  I think that's less confusing to the user: if ls shows a file with a weird name, s/he'll still be able to copy, rename or delete it.

# User-Facing Changes
Could break some existing scripts.  If user happened to have a file with a globbish name but was using a glob pattern with the same spelling, the new version will process the file and not expand the glob.

# Tests + Formatting

# After Submitting

### CODE REVIEW

**Comment by fdncred** [comment]
I just ran `toolkit check pr` and it succeeds on this PR but if you run `cargo test --workspace --profile ci --exclude nu_plugin_*`, which is what our ci runs, this PR fails on Windows. We need to get toolkit.nu inline with CI I think.

**Comment by bobhy** [comment]
Hoorah, tests all pass!
Please take a look and let me know what you think.

**Comment by fdncred** [comment]
I added some thoughts and context here https://github.com/nushell/nushell/issues/10364#issuecomment-1761407435

**Comment by amtoine** [comment]
@bobhy 
when you say

> Fixes #10571, also potentially: #10364, #10211, #9558, #9310 

could you be bit more precise?
these issues are not super straighforward to understand and i'm not sure if this PR should close them or not when i look at the new tests you've added :innocent:

**Comment by bobhy** [comment]
Well, I think the PR does close those issues, but it's up to the openers there to say.  I left a comment in each, inviting the opener to take a look

**Comment by amtoine** [comment]
> Well, I think the PR does close those issues, but it's up to the openers there to say. I left a comment in each, inviting the opener to take a look
>
ooh right, i missed them, my bad :pray:

**Comment by bobhy** [comment]
> I just ran `toolkit check pr` and it succeeds on this PR but if you run `cargo test --workspace --profile ci --exclude nu_plugin_*`, which is what our ci runs, this PR fails on Windows. We need to get toolkit.nu inline with CI I think.
> 
@fdncred 
I agree `toolkit.nu` should match CI tests. Is there something checked into the project that defines the CI tests?  I can't find it and am never sure I'm understanding what I see in CI run logs on github.

But, even it were perfectly aligned, `toolkit check pr` on linux is not going to detect a windows-only problem.

**Comment by fdncred** [comment]
@bobhy You need to look at the ci.yml. 

For clippy, I currently don't like how this is separated because it makes it hard to diagnose when there are problems, but if you look here 

https://github.com/nushell/nushell/blob/4e5a1ced131761a1b253080539ad29220e99942c/.github/workflows/ci.yml#L12

and then here

https://github.com/nushell/nushell/blob/4e5a1ced131761a1b253080539ad29220e99942c/.github/workflows/ci.yml#L51

For tests here

https://github.com/nushell/nushell/blob/4e5a1ced131761a1b253080539ad29220e99942c/.github/workflows/ci.yml#L55

Plugins have their own section.

As it relates to Windows, you can definitely use `toolkit check pr` on Windows, but maybe you mean that the checks run in the environment you're currently using?

**Comment by fdncred** [comment]
Thanks @bobhy. Anxious to see how people like this.

**Comment by bobhy** [comment]
Yah, me too.  We'll never hear from the folks for whom it just does the "right" thing.  I want to hear from the folks who say, "I typed `rm foo` and it deleted file foo, and that was the wrong thing to do"!

**Comment by fdncred** [comment]
The one thing that has always concerned me about this PR is `metadata`. Getting a file's metadata is notoriously slow. What concerns me is that now I think we're doing it multiple times. It's not like `ls` is super optimized but maybe something good to look at in the future is only calling the `.metadata()` method once per file for `ls`.

**Comment by bobhy** [comment]
OK, but it's small potatoes.  I'm doing 1 `stat`.  If it succeeds, that's the only one.  If  fails, all these commands are going to expand the glob and do at least 1 stat per file they find (in `nu_glob::fill_todo()` to check for directory).

File: crates/nu-cmd-base/src/arg_glob.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    match md {
+        Ok(_metadata) => {
+            return Ok(Paths::single(&PathBuf::from(pattern.item), cwd));
+        }
+        // file not found, but also "invalid chars in file" (e.g * on Windows).  Fall through and glob
+        Err(_) => {}
+    }
+
+    // user wasn't referring to a specific thing in filesystem, try to glob it.
+    match glob_with_parent(&pattern.item, options, cwd) {
```
Comment (WindSoilder) [inline crates/nu-cmd-base/src/arg_glob.rs:65]: Hi, @bobhy 
Sorry for a simple question, why are you using `glob_with_parent` rather than `glob_with`?

**Pull request merged**